firedrake/preconditioners/fdm.py

```diff
@@ -0,0 +1,1134 @@
+                    raise ValueError("I don't know what to do with the homogeneity tensor")
+        else:
+            # immersed manifold mesh
+            F = Jacobian(mesh)
+            if len(alpha.ufl_shape) == 2 and Piola is None:
+                G = inv(dot(dot(F.T, inv(alpha)), F))
+            else:
+                raise ValueError("I don't know what to do with the homogeneity tensor")
+
+        if discard_mixed:
+            # discard mixed derivatives and mixed components
+            if len(G.ufl_shape) == 2:
+                G = diag_vector(G)
+                Qe = VectorElement(family, mesh.ufl_cell(), degree=degree,
+                                   quad_scheme="default", dim=numpy.prod(G.ufl_shape))
+            elif len(G.ufl_shape) == 4:
+                G = as_tensor([[G[i, j, i, j] for j in range(G.ufl_shape[1])] for i in range(G.ufl_shape[0])])
+                Qe = TensorElement(family, mesh.ufl_cell(), degree=degree,
+                                   quad_scheme="default", shape=G.ufl_shape)
+            else:
+                raise ValueError("I don't know how to discard mixed entries of a tensor with shape ", G.ufl_shape)
+        else:
+            Qe = TensorElement(family, mesh.ufl_cell(), degree=degree,
+                               quad_scheme="default", shape=G.ufl_shape, symmetry=True)
+
+        # assemble second order coefficient
+        Q = firedrake.FunctionSpace(mesh, Qe)
+        q = firedrake.TestFunction(Q)
+        Gq = firedrake.Function(Q)
+        coefficients["Gq"] = Gq
+        assembly_callables.append(partial(firedrake.assemble, inner(G, q)*dx(degree=quad_deg), Gq))
+
+        # assemble zero-th order coefficient
+        if isinstance(beta, Zero):
+            Bq = None
+        else:
+            if Piola:
+                # apply Piola transform and keep diagonal
+                beta = diag_vector(dot(Piola.T, dot(beta, Piola)))
+            shape = beta.ufl_shape
+            if shape:
+                Qe = TensorElement(family, mesh.ufl_cell(), degree=degree,
+                                   quad_scheme="default", shape=shape)
+            else:
+                Qe = FiniteElement(family, mesh.ufl_cell(), degree=degree,
+                                   quad_scheme="default")
+
+            Q = firedrake.FunctionSpace(mesh, Qe)
+            q = firedrake.TestFunction(Q)
+            Bq = firedrake.Function(Q)
+            coefficients["Bq"] = Bq
+            assembly_callables.append(partial(firedrake.assemble, inner(beta, q)*dx(degree=quad_deg), Bq))
+
+        if needs_hdiv:
+            # make DGT functions with the second order coefficient
+            # and the Piola transform matrix for each side of each facet
+            extruded = mesh.cell_set._extruded
+            dS_int = firedrake.dS_h(degree=quad_deg) + firedrake.dS_v(degree=quad_deg) if extruded else firedrake.dS(degree=quad_deg)
+            ele = BrokenElement(FiniteElement("DGT", mesh.ufl_cell(), 0))
```

Comment:
So this is one dof per facet, but associated with the cell because you need different values on either side of the facet I guess.